include/deal.II/lac/trilinos_tpetra_vector.h

```diff
@@ -826,28 +954,44 @@ namespace LinearAlgebra
 
       for (size_type i = 0; i < n_elements; ++i)
         {
-          const size_type                         row = indices[i];
-          const TrilinosWrappers::types::int_type local_row =
+          const size_type                   row = indices[i];
+          TrilinosWrappers::types::int_type local_row =
             vector->getMap()->getLocalElement(row);
 
+          if (local_row == Teuchos::OrdinalTraits<int>::invalid())
+            {
+              Assert(compressed && !relevant_elements.is_null(),
+                     ExcAccessOffProcessorEntry());
+              if (compressed && !relevant_elements.is_null())
+                {
+                  // attempt to decompress the vector
+                  this->decompress();
```

Comment:
This cannot work: You can't call a collective operation inside a loop because processes will get here a different number of times (or not at all: think of a situation where a process has no locally owned cells, and so never participates in the assembly of vectors and matrices).

Reply:
You are right, I have not carefully thought through this...
And in my own tests, this case never occurred, so I have not noticed, that this will not work.

So the only thing we can do in that case is to throw an error.